text/0014-monitoring.md

```diff
@@ -0,0 +1,66 @@
+# Collect and Expose Metrics
+
+## Summary
+
+[summary]: #summary
+
+The goal is to collect and expose structured metrics about the quality and status of the SLAM algorithm and the server that runs it.
+
+## Motivation
+
+[motivation]: #motivation
+
+### Dashboard for humans
+
+Human developers debug SLAM performance, and human operators need to monitor key metrics for the robot.
+Both can profit greatly from a nice dashboard that shows key cartographer metrics, such as a histogram of constraint quality.
+
+Today, developers often find themselves searching in log output, and the only way to monitor Cartographer is to run rviz at runtime.
+We need to provide a more structured way that also works for multiple instances or longer time scales.
+
+### Reference signal for robot control
+
+Some manufacturers would like to have an observable for SLAM performance so that they can make downstream decisions due to bad SLAM.
+Collected metrics could provide these reference signals with little intrusion in Cartographer code.
+
+## Approach
+
+[approach]: #approach
+
+The approach below is illustrated by a [prototype](https://github.com/googlecartographer/cartographer/pull/876).
+
+### Cartographer facade
+
+Inside the cartographer library, we establish a system-agnostic facade for collecting metrics.
+The facade is compatible with all important monitoring systems (Prometheus, Stackdriver, OpenCensus), but doesn't introduce build dependencies into the core cartographer lib.
+The facade lives in the cartographer::metrics namespace and includes these core classes:
+
+*   `Counter`: A single, always-increasing count of events, e.g. "number of loop iterations".
+*   `Gauge`: A value that changes both up and down, e.g. "memory used" or "open requests".
+*   `Histogram`: A thread-safe histogram with predefined buckets and constant memory usage.
+    It's different from `cartographer::common::Histogram`, which is thread-unsafe and does the bucketing after value collection.
+    Over time, I think that there would be potential to merge the two Histogram classes.
+*   `CounterFamily`, `GaugeFamily`, `HistogramFamily`: An interface for obtaining `Counter`/`Gauge`/`Histogram` instances that share a common purpose and only differ by a set of labels.
```

Comment:
You could add your explanation from the comment "Metrics in the same family can safely be aggregated."

Reply:
Done.